File: orchestrator/vcs/gitlab.py
import json
import os
import subprocess


_CI_STATE_MAP = {
    "success": "success",
    "failed": "failure",
    "canceled": "failure",
    "skipped": "success",
    "running": "pending",
    "pending": "pending",
    "created": "pending",
    "waiting_for_resource": "pending",
    "preparing": "pending",
    "scheduled": "pending",
    "manual": "pending",
}

_MAX_LOG_CHARS = 8000


class GitLabProvider:
    def __init__(self, host: str = "gitlab.com") -> None:
        self.host = host

    def _env(self) -> dict:
        env = os.environ.copy()
        env["GITLAB_HOST"] = self.host
        return env

    def create_pr(
        self,
        task_id: str,
        title: str,
        body: str,
        head: str,
        base: str,
    ) -> str | None:
        result = subprocess.run(
            [
                "glab", "mr", "create",
                "--title", title,
                "--description", body,
                "--source-branch", head,
                "--target-branch", base,
                "--yes",
            ],
            capture_output=True,
            text=True,
            env=self._env(),
        )
        if result.returncode != 0:
            return None
        # glab prints the MR URL on the last non-empty line
        lines = [l.strip() for l in result.stdout.splitlines() if l.strip()]
        return lines[-1] if lines else None

    def get_ci_status(self, pr_url: str) -> str:
        # Derive MR IID from URL (/merge_requests/<iid>)
        iid = self._iid_from_url(pr_url)
        if not iid:
            return "pending"

        result = subprocess.run(
            ["glab", "mr", "view", iid, "--output", "json"],
            capture_output=True,
            text=True,
            env=self._env(),
        )
        if result.returncode != 0:
            return "pending"

        try:
            data = json.loads(result.stdout)
        except (json.JSONDecodeError, ValueError):
            return "pending"

        pipeline_status = data.get("head_pipeline", {}).get("status", "pending")
        return _CI_STATE_MAP.get(pipeline_status, "pending")

    def get_ci_logs(self, pr_url: str) -> str:
        iid = self._iid_from_url(pr_url)
        if not iid:
            return ""

        # Get pipeline ID
        view_result = subprocess.run(
            ["glab", "mr", "view", iid, "--output", "json"],
            capture_output=True,
            text=True,
            env=self._env(),
        )
        if view_result.returncode != 0:
            return ""

        try:
            data = json.loads(view_result.stdout)
            pipeline_id = str(data.get("head_pipeline", {}).get("id", ""))
        except (json.JSONDecodeError, ValueError):
            return ""

        if not pipeline_id:
            return ""

        # Get failed job IDs
        jobs_result = subprocess.run(
            ["glab", "pipeline", "jobs", pipeline_id, "--output", "json"],
            capture_output=True,
            text=True,
            env=self._env(),
        )
        if jobs_result.returncode != 0:
            return ""

        try:
            jobs = json.loads(jobs_result.stdout)
            failed_ids = [str(j["id"]) for j in jobs if j.get("status") == "failed"]
        except (json.JSONDecodeError, ValueError, KeyError):
            return ""

        logs: list[str] = []
        for job_id in failed_ids[:3]:  # limit to first 3 failed jobs
            trace_result = subprocess.run(
                ["glab", "pipeline", "trace", job_id],
                capture_output=True,
                text=True,
                env=self._env(),
            )
            if trace_result.returncode == 0:
                logs.append(trace_result.stdout)

        combined = "\n---\n".join(logs)
        return combined[:_MAX_LOG_CHARS]

    def merge_pr(self, pr_url: str) -> bool:
        iid = self._iid_from_url(pr_url)
        if not iid:
            return False

        result = subprocess.run(
            ["glab", "mr", "merge", iid, "--squash", "--yes"],
            capture_output=True,
            text=True,
            env=self._env(),
        )
        return result.returncode == 0

    @staticmethod
    def _iid_from_url(url: str) -> str:
        parts = url.rstrip("/").split("/")
        idx = next((i for i, p in enumerate(parts) if p == "merge_requests"), -1)
        if idx != -1 and idx + 1 < len(parts):
            return parts[idx + 1]
        return ""

File: internal/vcs/gitlab.go
package vcs

import (
	"encoding/json"
	"strconv"
	"strings"
)

// gitlabPipelineStatusMap ports _CI_STATE_MAP from orchestrator/vcs/gitlab.py.
var gitlabPipelineStatusMap = map[string]CIState{
	"success":              CISuccess,
	"failed":               CIFailure,
	"canceled":             CIFailure,
	"skipped":              CISuccess,
	"running":              CIPending,
	"pending":              CIPending,
	"created":              CIPending,
	"waiting_for_resource": CIPending,
	"preparing":            CIPending,
	"scheduled":            CIPending,
	"manual":               CIPending,
}

// GitLabProvider drives MR creation, CI polling, and merges through the
// `glab` CLI. Ports GitLabProvider (orchestrator/vcs/gitlab.py).
type GitLabProvider struct {
	host string
}

// NewGitLabProvider builds a GitLabProvider talking to host (e.g.
// "gitlab.com" or a self-hosted instance).
func NewGitLabProvider(host string) *GitLabProvider {
	return &GitLabProvider{host: host}
}

func (p *GitLabProvider) env() []string {
	return []string{"GITLAB_HOST=" + p.host}
}

func (p *GitLabProvider) CreatePR(head, base, title, body string) (string, error) {
	if err := checkBinary("glab"); err != nil {
		return "", err
	}
	out, err := run("glab", p.env(),
		"mr", "create",
		"--title", title,
		"--description", body,
		"--source-branch", head,
		"--target-branch", base,
		"--yes",
	)
	if err != nil {
		if authErr := checkAuth("glab", p.env()); authErr != nil {
			return "", authErr
		}
		return "", nil
	}
	url := lastNonEmptyLine(out)
	if url == "" {
		return "", nil
	}
	return url, nil
}

type gitlabMRView struct {
	HasConflicts bool `json:"has_conflicts"`
	HeadPipeline struct {
		ID     json.Number `json:"id"`
		Status string      `json:"status"`
	} `json:"head_pipeline"`
}

func (p *GitLabProvider) CIStatus(prURL string) (CIState, error) {
	if err := checkBinary("glab"); err != nil {
		return "", err
	}
	iid := iidFromURL(prURL)
	if iid == "" {
		return CIPending, nil
	}

	out, err := run("glab", p.env(), "mr", "view", iid, "--output", "json")
	if err != nil {
		return CIPending, nil
	}

	var view gitlabMRView
	if err := json.Unmarshal([]byte(out), &view); err != nil {
		return CIPending, nil
	}
	if view.HeadPipeline.Status == "" {
		// No pipeline at all: see CINone and CIConflict.
		if view.HasConflicts {
			return CIConflict, nil
		}
		return CINone, nil
	}
	if mapped, ok := gitlabPipelineStatusMap[view.HeadPipeline.Status]; ok {
		return mapped, nil
	}
	return CIPending, nil
}

type gitlabJob struct {
	ID     json.Number `json:"id"`
	Status string      `json:"status"`
}

func (p *GitLabProvider) CILogs(prURL string) (string, error) {
	if err := checkBinary("glab"); err != nil {
		return "", err
	}
	iid := iidFromURL(prURL)
	if iid == "" {
		return "", nil
	}

	viewOut, err := run("glab", p.env(), "mr", "view", iid, "--output", "json")
	if err != nil {
		return "", nil
	}
	var view gitlabMRView
	if err := json.Unmarshal([]byte(viewOut), &view); err != nil {
		return "", nil
	}
	pipelineID := view.HeadPipeline.ID.String()
	if pipelineID == "" {
		return "", nil
	}

	jobsOut, err := run("glab", p.env(), "pipeline", "jobs", pipelineID, "--output", "json")
	if err != nil {
		return "", nil
	}
	var jobs []gitlabJob
	if err := json.Unmarshal([]byte(jobsOut), &jobs); err != nil {
		return "", nil
	}

	var failedIDs []string
	for _, j := range jobs {
		if j.Status == "failed" {
			failedIDs = append(failedIDs, j.ID.String())
		}
	}
	if len(failedIDs) > 3 {
		failedIDs = failedIDs[:3]
	}

	var logs []string
	for _, jobID := range failedIDs {
		traceOut, err := run("glab", p.env(), "pipeline", "trace", jobID)
		if err == nil {
			logs = append(logs, traceOut)
		}
	}
	combined := strings.Join(logs, "\n---\n")
	return truncate(combined, maxLogChars), nil
}

func (p *GitLabProvider) MergePR(prURL string, squash, auto bool) error {
	if err := checkBinary("glab"); err != nil {
		return err
	}
	iid := iidFromURL(prURL)
	if iid == "" {
		return ErrMergeFailed
	}

	args := []string{"mr", "merge", iid, "--yes"}
	if squash {
		args = append(args, "--squash")
	}
	if auto {
		args = append(args, "--when-pipeline-succeeds")
	}
	if _, err := run("glab", p.env(), args...); err != nil {
		if authErr := checkAuth("glab", p.env()); authErr != nil {
			return authErr
		}
		return ErrMergeFailed
	}
	return nil
}

// iidFromURL extracts the numeric MR IID from ".../merge_requests/<iid>".
// Ports GitLabProvider._iid_from_url.
func iidFromURL(url string) string {
	parts := strings.Split(strings.TrimRight(url, "/"), "/")
	for i, part := range parts {
		if part == "merge_requests" && i+1 < len(parts) {
			candidate := parts[i+1]
			if _, err := strconv.Atoi(candidate); err == nil {
				return candidate
			}
			return ""
		}
	}
	return ""
}
